File: ch7/mirror.py
#simulation of magnetic mirror

import numpy as np
import pylab as pl


def evalB(z0):
    z = z0/1e-5
    B = np.exp(-2*(z-1)**2) + np.exp(-2*(z+1)**2)
    return B

def evalGradB(z0):
    z = z0/1e-5
    gB = np.exp(-2*(z-1)**2)*(-4*(z-1)) + np.exp(-2*(z+1)**2)*(-4*(z+1))
    return gB*1e5  #because of unit conversion


vz=1e1

q = -1.602e-19
m = 9.19e-31	#electron mass

#electric field
E = np.asarray([0,0,0])

#initial position and velocity
max_it=20000
x=np.zeros((max_it,3))
v=np.zeros((max_it,3))

Bmag = np.zeros(max_it)
x[0,:] = [0,0,0]
v[0,:] = [1e4,0,1.6e4]
dt0 = 1e-12

#compute some values
Rm = (1+np.exp(-8))/(2*np.exp(-2))	#Bm/B0
theta = np.arcsin(np.sqrt(1/Rm))*180/np.pi
vperp2 = v[0,0]**2+v[0,1]**2
vpar2 = v[0,2]**2
print("theta: %.3g,   1/Rm: %.3g, vp2/v02: %.3g"%(theta,1/Rm, vperp2/(vperp2+vpar2)))

B0 = evalB(0);
vtot2 = vperp2+vpar2
print("Particle mirrors at Bm: %.3g, vtot:%.3g"%(B0*vtot2/vperp2, np.sqrt(vtot2)))
#magnetic moment, should be conserved
mu0 = vperp2/evalB(x[0,2])
mu = 0.5*m*mu0
Bmag[0] = vperp2/mu0	

for it in range (max_it-1):

	#evaluate B at particle position
	z = x[it,2]
	B = np.asarray([0,0,evalB(z)])
	gradB = np.asarray([0,0,evalGradB(z)])
	
	#force
	F=q*E - mu*gradB
    	
     #velocity rewind for leapfrog
	if (it==0):
         dt = 0.5*dt0
	else:
         dt = dt0
         
	#v minus
	v_minus = v[it,:] + (F/m)*0.5*dt
	
	#v prime
	t = q/m*B*0.5*dt	#t vector
	v_prime = v_minus + np.cross(v_minus,t)
	
	#v plus
	s = 2*t/(1+np.dot(t,t))#    #s vector
	v_plus = v_minus + np.cross(v_prime,s)
	
	#v n+1/2
	v[it+1,:] = v_plus + (F/m)*0.5*dt
	
	#position update
	x[it+1,:] = x[it,:] + v[it+1,:]*dt0
	
	#save bmag for plotting
	Bmag[it+1] = B[2]
	
	#screen output
	if (np.mod(it,500)==0):
		#perpendicular and tangential velocities
		vperp2 = v[it,0]**2 + v[it,1]**2
		vpar2 = v[it,2]**2
		
		print("it:%d, z:%.2e, v_par:%.2e, v_perp:%.2e, v_tot:%.4e"%
		(it,x[it,2],np.sqrt(vperp2),np.sqrt(vpar2),np.sqrt(vperp2+vpar2)))
	
pl.figure(1)
pl.plot (x[:,0],x[:,1])
pl.title("x-y view")
pl.axis('equal')

pl.figure(2)
pl.plot (np.arange(0,max_it),x[:,2])
pl.title("it vs. z")

pl.figure(3)
pl.plot (np.arange(0,max_it),(v[:,0]**2 + v[:,1]**2)/B[2])
pl.title("it vs. mu0")

pl.figure(4)
pl.plot(x[:,2],Bmag)
pl.title("z vs Bmag")

File: ch7/mirror.cpp
//demo of particle integration with magnetic field with the Boris method
#include <fstream>
#include <iostream>
#include "Vec.h"

using namespace std;

//simple particle container
struct Particle {
	double3 x,v;	//position and velocity
	double q;		//charge
	double m;		//mass
	double mu = -1;		//
	Particle(const double3 &x, const double3 &v, double q, double m):
	x{x}, v{v}, q{q}, m{m} {}
};

double3 evalB(double z_in) {
    double z = z_in/1e-5;
    double Bz = exp(-2*(z-1)*(z-1)) + exp(-2*(z+1)*(z+1));
    return {0,0,Bz};
}

double3 evalGradB(double z_in) {
    double z = z_in/1e-5;
    double gBz = exp(-2*(z-1)*(z-1))*(-4*(z-1)) + exp(-2*(z+1)*(z+1))*(-4*(z+1));
    return {0,0,gBz*1e5};  //because of unit conversion
}

//Boris method
double updateVelocityBoris(Particle &part, double dt, const double3 &E)
{
	double qm = part.q/part.m;		//q/m

	//evaluate B field and gradB at particle position
	double3 B = evalB(part.x[2]);
	double3 gradB = evalGradB(part.x[2]);

	//unit vector in B direction
	double3 B_unit = unit(B);
	

	double mag1, mag2;
	//magnetic moment, should be conserved
	if (part.mu<0) {
		double3 v_par = dot(part.v,B_unit)*B_unit;
		double3 v_perp = part.v - v_par;	
		part.mu=0.5*part.m*dot(v_perp,v_perp)/mag(B);	//initialize
	}		
	else {
	//orbital velocity, part.v - v_parallel	
		mag1 = mag(part.v);	
		
		double3 v_par = dot(part.v,B_unit)*B_unit;
		double3 v_perp = part.v - v_par;
		double v_perp_mag_new = sqrt(part.mu/part.m*2*mag(B));
		double v_perp_mag = mag(v_perp);
		v_perp = v_perp_mag_new*unit(v_perp);
		double v_par2 =mag1*mag1 - v_perp_mag_new*v_perp_mag_new;
		double v_par_mag_new;		
		if (v_par2>=0) v_par_mag_new = sqrt(v_par2);
		else v_par_mag_new = -sqrt(-v_par2);
		v_par = v_par_mag_new * unit(v_par);
		part.v = v_perp+v_par;	
		mag2 = mag(part.v);

	}

	//total acceleration force
	double3 F = part.q*E; //- mu*gradB*unit(gradB);
	
    //v minus, first half acceleration
	double3 v_minus = part.v + (F/part.m)*dt/2;

	//v prime, first half rotation
	double3 t = qm*B*dt/2;
	double3 v_prime = v_minus + cross(v_minus, t);

	//v_plus, second half rotation
	double3 s = 2.0*t/(1+dot(t,t));
	double3 v_plus = v_minus + cross(v_prime, s);

	//v[k+0.5], second half acceleration
	part.v = v_plus + (F/part.m)*dt/2;
//cout<<part.x[2]<<": "<<sqrt(part.v[0]*part.v[0]+part.v[1]*part.v[1]);

//cout<<" ZZ "<<part.v[2]<<" ("<<mag1<<", "<<mag2<<")"<<endl;
		double3 v_par = dot(part.v,B_unit)*B_unit;
		double3 v_perp = part.v - v_par;
	return 0.5*part.m*dot(v_perp,v_perp)/mag(B);
}

int main() {

	double3 E{0,0,0};		//no electric field

	//compute Larmor radius for an electron
	double q = -Const::QE;
	double m = Const::ME;
	double vt = 1e4;
	double3 B = evalB(0);
	double rL = m*vt/(abs(q)*mag(B));
	
	//create two electrons with orbit centered on origin
	Particle part({rL,0,0},{0,vt,1.6e4},q,m);
	double dt = 1e-12;
	
	//output file
	ofstream out("trace.csv");
	out<<"ts x y z u v w v_perp v_par mag(v) Bz gradBz mu\n";

	double mu;

	//rewind velocity per leapfrog
	mu = updateVelocityBoris(part,-0.5*dt,E);
	
	//main loop
	for (int ts = 0;ts<=20000;ts++) {		
		//output data
		out<<ts<<" "<<part.x<<" "<<part.v<<" "<<sqrt(part.v[0]*part.v[0]+part.v[1]*part.v[1])<<" "<<part.v[2]<<" " <<mag(part.v)<<" "<<evalB(part.x[2])[2]<<" "<<evalGradB(part.x[2])[2]<<" "<<mu<<"\n";		

		mu = updateVelocityBoris(part,dt,E);
		part.x += part.v*dt;
	}	

return 0;
} 

